controllers/datasource_controller.go

```diff
@@ -87,38 +101,67 @@ func (r *DatasourceReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Complete(r)
 }
 
-// Checkdatasource to update status
-func (r *DatasourceReconciler) Checkdatasource(ctx context.Context, logger logr.Logger, instance *arcadiav1alpha1.Datasource) error {
-	logger.Info("check datasource")
-	secret := corev1.Secret{}
-	if err := r.Get(ctx, types.NamespacedName{Namespace: instance.Namespace, Name: instance.Spec.AuthSecret}, &secret); err != nil {
-		return err
+func (r *DatasourceReconciler) Initialize(ctx context.Context, logger logr.Logger, instance *arcadiav1alpha1.Datasource) error {
+	instanceDeepCopy := instance.DeepCopy()
+	l := len(instanceDeepCopy.Finalizers)
+
+	var update bool
+
+	instanceDeepCopy.Finalizers = utils.AddString(instanceDeepCopy.Finalizers, arcadiav1alpha1.Finalizer)
+	if l != len(instanceDeepCopy.Finalizers) {
+		logger.V(1).Info("Add Finalizer for datasource", "Finalizer", arcadiav1alpha1.Finalizer)
+		update = true
 	}
-	accessKeyID := string(secret.Data[rootUser])
-	secretAccessKey := string(secret.Data[rootPassword])
-	endpoint := instance.Spec.URL
 
-	useSSL := false
+	if instanceDeepCopy.Labels == nil {
+		instanceDeepCopy.Labels = make(map[string]string)
+	}
 
-	// Initialize minio client object.
-	minioClient, err := minio.New(endpoint, &minio.Options{
-		Creds:  credentials.NewStaticV4(accessKeyID, secretAccessKey, ""),
-		Secure: useSSL,
-	})
-	if err != nil {
-		return err
+	currentType := string(instanceDeepCopy.Spec.Type())
+	if v := instanceDeepCopy.Labels[arcadiav1alpha1.LabelDatasourceType]; v != currentType {
+		instanceDeepCopy.Labels[arcadiav1alpha1.LabelDatasourceType] = currentType
+		update = true
 	}
-	buckets, err := minioClient.ListBuckets(ctx)
-	if err != nil {
-		return err
+
+	if update {
+		return r.Client.Update(ctx, instanceDeepCopy)
+	}
+
+	return nil
```

Comment:
**Suggestion:** The `Initialize` method is modifying the `instance` object but it's not clear if these changes are persisted. If the intention is to update the `instance` object in the Kubernetes API, it would be better to make this explicit by calling `r.Client.Update` after modifying the object.
```suggestion
func (r *DatasourceReconciler) Initialize(ctx context.Context, logger logr.Logger, instance *arcadiav1alpha1.Datasource) error {
  instanceDeepCopy := instance.DeepCopy()
  l := len(instanceDeepCopy.Finalizers)

  var update bool

  instanceDeepCopy.Finalizers = utils.AddString(instanceDeepCopy.Finalizers, arcadiav1alpha1.Finalizer)
  if l != len(instanceDeepCopy.Finalizers) {
    logger.V(1).Info("Add Finalizer for datasource", "Finalizer", arcadiav1alpha1.Finalizer)
    update = true
  }

  if instanceDeepCopy
```